Sources/Basic/ArgumentParser.swift

```diff
@@ -0,0 +1,223 @@
+        return name.hashValue
+    }
+
+    public static func ==(_ lhs: CommandLineArgument, _ rhs: CommandLineArgument) -> Bool {
+        return lhs.name == rhs.name
+    }
+}
+
+/// A class representing option arguments. These are optional arguments
+/// which may or may not be provided in the command line. They are always
+/// prefixed by their name. For e.g. --verbosity true.
+public class OptionArgument: CommandLineArgument {
+    override init(name: String, kind: Kind, usage: String?) {
+        precondition(!name.isPositionalArg)
+        super.init(name: name, kind: kind, usage: usage)
+    }
+}
+
+/// A class representing positional arguments. These arguments must be present
+/// and in the same order as they are added in the parser.
+public class PositionalArgument: CommandLineArgument {
+    override init(name: String, kind: Kind, usage: String?) {
+        precondition(name.isPositionalArg)
+        precondition(kind != .stringArray && kind != .bool)
+        super.init(name: name, kind: kind, usage: usage)
+    }
+}
+
+/// Argument parser struct reponsible to parse the provided array of arguments and return
+/// the parsed result.
+public struct ArgumentParser {
+    /// A struct representing result of the parsed arguments.
+    public struct ArgumentParserResult {
+        /// Internal representation of arguments mapped to their values.
+        private var results = [CommandLineArgument: Any]()
+
+        /// Adds a result.
+        ///
+        /// - parameter
+        ///     - argument: The argument for which this result is being added.
+        ///     - value: The associated value of the argument.
+        ///
+        /// - throws: ArgumentParserError
+        fileprivate mutating func addResult(for argument: CommandLineArgument, value: String) throws {
+            let result: Any
+
+            switch argument.kind {
+            case .bool:
+                result = true
+            case .int:
+                // Throw error if we're not able to convert to the type specified.
+                guard let intValue = Int(value) else {
+                    throw ArgumentParserError.typeMismatch("\(value) is not convertible to Int")
+                }
+                result = intValue
+            case .string:
+                result = value
+            case .stringArray:
+                var array = [String]()
+                // Get the previously added results if present.
+                if let previousResult = results[argument] as? [String] {
+                    array = previousResult
+                }
+                array.append(value)
+                result = array
+            }
+            // Save the result.
+            results[argument] = result
+        }
+
+        /// Get an option argument's value from the results.
+        ///
+        /// Since the options are optional, their result may or may not
+        /// be present.
+        public func get<T>(_ arg: OptionArgument) -> T? {
```

Comment:
You can do better here with stronger typing of the OptionArgument types.

If OptionArgument was a generic class, then you could do
```
public func get<T>(_ arg: OptionArgument<T>) -> T? {
  ...
}
```
and clients would get type safety of their declared arguments.